tests/test_client.py

```diff
@@ -108,5 +120,157 @@ async def launch_client() -> None:
     async with anyio.create_task_group() as tg:
         tg.start_soon(launch_client)
         await event.wait()
-        async with Client("test.mosquitto.org", will=Will(topic)) as client:
+        async with Client(HOSTNAME, will=Will(topic)) as client:
             client._client._sock_close()  # type: ignore[attr-defined]
+
+
+async def test_client_tls_context() -> None:
+    topic = TOPIC_HEADER + "tls_context"
+
+    async def handle_messages(tg: anyio.abc.TaskGroup) -> None:
+        async with client.filtered_messages(topic) as messages:
+            async for message in messages:
+                assert message.topic == topic
+                tg.cancel_scope.cancel()
+
+    async with Client(
+        HOSTNAME,
+        8883,
+        tls_context=ssl.SSLContext(protocol=ssl.PROTOCOL_TLS),
+    ) as client:
+        async with anyio.create_task_group() as tg:
+            await client.subscribe(topic)
+            tg.start_soon(handle_messages, tg)
+            await client.publish(topic)
+
+
+async def test_client_tls_params() -> None:
+    topic = TOPIC_HEADER + "tls_params"
+
+    async def handle_messages(tg: anyio.abc.TaskGroup) -> None:
+        async with client.filtered_messages(topic) as messages:
+            async for message in messages:
+                assert message.topic == topic
+                tg.cancel_scope.cancel()
+
+    async with Client(
+        HOSTNAME,
+        8883,
+        tls_params=TLSParameters(
+            ca_certs=str(Path.cwd() / "tests" / "mosquitto.org.crt")
+        ),
+    ) as client:
+        async with anyio.create_task_group() as tg:
+            await client.subscribe(topic)
+            tg.start_soon(handle_messages, tg)
+            await client.publish(topic)
+
+
+async def test_client_username_password() -> None:
+    topic = TOPIC_HEADER + "username_password"
+
+    async def handle_messages(tg: anyio.abc.TaskGroup) -> None:
+        async with client.filtered_messages(topic) as messages:
+            async for message in messages:
+                assert message.topic == topic
+                tg.cancel_scope.cancel()
+
+    async with Client(HOSTNAME, username="asyncio-mqtt", password="012") as client:
+        async with anyio.create_task_group() as tg:
+            await client.subscribe(topic)
+            tg.start_soon(handle_messages, tg)
+            await client.publish(topic)
+
+
+async def test_client_logger() -> None:
+    logger = logging.getLogger("asyncio-mqtt")
+    async with Client(HOSTNAME, logger=logger) as client:
+        assert logger == client._client._logger  # type: ignore[attr-defined]
+
+
+async def test_client_max_concurrent_outgoing_calls(
```

Comment:
Now this is an interesting test! Well done. 👍

I wonder how many people actually use this feature in production. 😄 

This also showcases a weakness in our current API: We don't have any backpressure mechanics. *That* is difficult to solve "right", though. Definitely a 2.0 kind of thing (or with the anyio rewrite/replacement).

If you're interested in that kind of stuff there is a good discussion here: https://github.com/python-trio/trio/issues/987

Reply:
Is not `max_concurrent_outgoing_calls` a backpressure mechanism?

Reply:
It is but it is a very crude one at that. 😄 My point is that we can and should do better in a future API.